pallets/handles/src/handles-utils/src/converter.rs

```diff
@@ -0,0 +1,119 @@
+//! # Handle Converter
+//!
+//! `handle_converter` provides a `HandleConverter` struct to detect confusable Unicode characters in a
+//! given input string and return its canonical form.
+use crate::{confusables::CONFUSABLES, types::*};
+use unicode_normalization::{char::is_combining_mark, UnicodeNormalization};
+extern crate alloc;
+use alloc::{
+	string::{String, ToString},
+	vec::Vec,
+};
+
+/// A converter for confusable characters.
+///
+/// Given a string, detects easily confusable characters and returns the string in canonical form.
+pub struct HandleConverter {}
+/// Creates a new `HandleConverter` instance with the specified input string.
+impl HandleConverter {
+	/// Creates a new `HandleConverter` instance with a built confusables map.
+	/// Converts a given string to its canonical form by stripping Unicode whitespace,
+	/// replacing confusable characters, and stripping diacritical marks.
+	/// The resulting string is converted to lowercase ASCII characters.
+	///
+	/// # Arguments
+	///
+	/// * `input_str` - The input string to convert to canonical form.
+	///
+	/// # Returns
+	///
+	/// A new string in canonical form.
+	///
+	pub fn convert_to_canonical(input_str: &str) -> alloc::string::String {
+		let white_space_stripped = Self::strip_unicode_whitespace(input_str);
+		let confusables_removed = Self::replace_confusables(&white_space_stripped);
+		let diacriticals_stripped = Self::strip_diacriticals(&confusables_removed);
+		diacriticals_stripped.to_ascii_lowercase()
+	}
+
+	/// Replaces any characters in the input string that are confusable with a different character.
+	///
+	/// # Arguments
+	///
+	/// * `input_str` - A reference to the input string to replace confusable characters.
+	///
+	/// # Returns
+	///
+	/// A new string where any confusable characters have been replaced with their corresponding non-confusable characters.
+	pub fn replace_confusables(input_str: &str) -> alloc::string::String {
+		input_str
+			.chars()
+			.map(|character| CONFUSABLES.get(&character).map_or(character, |&value| value))
+			.collect::<alloc::string::String>()
+	}
+
+	/// This function removes diacritical marks from the input string and returns a new `String` without them.
+	///
+	/// # Arguments
+	/// * `input_str` - A string slice that contains the input string from which the diacritical marks need to be removed.
+	///
+	/// # Notes
+	/// This function uses the NFKD normalization form and filtering of combining marks to strip the diacritical marks from the input string.
+	/// Combining marks are Unicode characters that are intended to modify the appearance of another character.
+	///
+	pub fn strip_diacriticals(input_str: &str) -> alloc::string::String {
```

Comment:
maybe this does not need to be public?

Reply:
It depends on the purpose of the package/crate.  If it's meant to be an external API, it should remain.  If it's just to modularize functions for the pallet, then it could be either private/public.